File: euphony/src/main/java/co/euphony/common/Packet.java
package co.euphony.common;

import java.util.Stack;

import co.euphony.util.PacketErrorDetector;

public class Packet {

    int[] payload;
    int checksum;
    int parityCode;
    boolean isVerified;

    Stack<Integer> packetStack = new Stack<>();

    public Packet() {
        clear();
    }

    public void push(int data) {
        packetStack.push(data);
    }

    public void clear() {
        packetStack.clear();
        payload = null;
        checksum = -1;
        parityCode = -1;
        isVerified = false;
    }

    public boolean build() {
        parityCode = packetStack.pop();
        checksum = packetStack.pop();

        payload = new int[packetStack.size()];
        for(int i = payload.length - 1; i >= 0; i--) {
            payload[i] = packetStack.pop();
        }

        boolean isChecksumOK = PacketErrorDetector.verifyCheckSum(payload, checksum);
        boolean isParityCodeOK = PacketErrorDetector.verifyEvenParity(payload, parityCode);

        isVerified = isChecksumOK && isParityCodeOK;
        return isVerified;
    }

    public boolean isVerified() {
        return isVerified;
    }

    public int[] getPayload() {
        return payload;
    }

    public String toString() {
        StringBuilder ret = new StringBuilder();
        ret.append("S");
        for(int val : payload) {
            ret.append(getInt2HexString(val));
        }
        ret.append(getInt2HexString(checksum));
        ret.append(getInt2HexString(parityCode));

        return ret.toString();
    }

    private String getInt2HexString(int val) {
        if(isMoreThan0xA(val))
            return String.valueOf((char)('a' + (val - 10)));
        else
            return String.valueOf(val);
    }

    private boolean isMoreThan0xA(int hex) { return hex > 9; }
}

File: euphony/src/main/cpp/core/source/Packet.cpp

#include "../Packet.h"
#include "../PacketErrorDetector.h"
#include <sstream>
using namespace Euphony;

string Packet::create(string source) {
    std::stringstream result;

    string encodedCode = base16.encode(source);
    for(char& c : encodedCode) {
        payload.push_back(base16.convertChar2Int(c));
    }

    string errorCode = makeErrorDetectorCode();

    result << "S" << encodedCode << errorCode;
    return result.str();
}

void Packet::push(int data) {
    payload.push_back(data);
}

void Packet::clear() {
    this->checksum = -1;
    this->parityCode = -1;
    this->isVerified = false;
    (this->payload).clear();
}

string Packet::makeErrorDetectorCode() {
    string errorCode = PacketErrorDetector::makeParityAndChecksum(payload);
    this->checksum = Base16::convertChar2Int(errorCode.at(0));
    this->parityCode = Base16::convertChar2Int(errorCode.at(1));
    this->isVerified = true;

    return errorCode;
}


string Packet::getStrPayload() {
    std::stringstream result;

    for(auto data : payload) {
        result << std::hex << data;
    }

    return result.str();
}

string Packet::toString() {
    std::stringstream result;

    result << "S";
    for(auto data : payload) {
        result << std::hex << data;
    }
    result << makeErrorDetectorCode();

    return result.str();
}



